byterun/debugger.c

```diff
@@ -194,11 +195,16 @@ void caml_debugger_init(void)
     /* Unix domain */
     sock_domain = PF_UNIX;
     sock_addr.s_unix.sun_family = AF_UNIX;
+    a_len = strlen(address);
+    if (a_len >= sizeof(sock_addr.s_unix.sun_path)) {
+      caml_fatal_error("Debug socket path length exceeds maximum permitted length");
+    }
     strncpy(sock_addr.s_unix.sun_path, address,
-            sizeof(sock_addr.s_unix.sun_path));
+            sizeof(sock_addr.s_unix.sun_path) - 1);
+    sock_addr.s_unix.sun_path[sizeof(sock_addr.s_unix.sun_path) - 1] = '\0';
     sock_addr_len =
       ((char *)&(sock_addr.s_unix.sun_path) - (char *)&(sock_addr.s_unix))
-        + strlen(address);
+        + a_len;
 #else
```

Comment:
`man 7 unix` on Linux suggests to simply use `sizeof(struct  sockaddr_un)` as the length parameter.  Maybe the original code was just wrong trying to track the actual length of the path.

Reply:
On Linux yes, but it wasn't clear to me what the behaviour is elsewhere. That man page says:

```
The addrlen argument that describes the enclosing sockaddr_un
          structure should have a value of at least:

              offsetof(struct sockaddr_un, sun_path)+strlen(addr.sun_path)+1

          or, more simply, addrlen can be specified as sizeof(struct sock‐
          addr_un).
```

which led me to conclude that the existing length calculation is more precise and possibly what used to happen before it became acceptable to just pass the whole (presumably NUL padded) struct.

I think I'd prefer to merge this minimal patch to fix the overflow, and then peer more closely at the use of globals and various socket options and perhaps add some test cases in a future post-4.07 PR :-)

Reply:
I fully agree with your conclusions.  I'm merging in trunk and in 4.07 right now.